dali/kernels/slice/slice_gpu.cuh

```diff
@@ -72,9 +72,28 @@ struct SliceBlockDesc {
   uint64_t size;
 };
 
-template<typename OutputType>
-constexpr int coalesced_values = sizeof(OutputType) >= 4 ? 1 : 4 / sizeof(OutputType);
-
+template<typename T>
+union PackedBuffer {
+  using PackedType = uint32_t;
+  static constexpr size_t kCapacity = sizeof(T) >= sizeof(PackedType) ?
+                                      1 : sizeof(PackedType) / sizeof(T);
+
+  T values[kCapacity];
+  PackedType raw;
+
+  __device__ inline void store(T* mem, size_t count) {
+    if (kCapacity == 1) {
+      *mem = *values;
+    } else if (count == kCapacity && (uint64_t)mem % sizeof(PackedType) == 0) {
+      *reinterpret_cast<PackedType*>(mem) = raw;
```

Comment:
In general this is UB, what would be the difference if you use `memcpy` instead here (which is the only recommended way according to the standard).

Reply:
I didn't know that was UB, thank you! I'll post benchmarks with `memcpy` in the evening.

Reply:
Bad news – `memcpy` compiles to (unrolled) byte-by-byte copy:
```
 ST.E.U8.SYS [R12], R10 
 ST.E.U8.SYS [R12+0x1], R9 
 ST.E.U8.SYS [R12+0x2], R6 
 ST.E.U8.SYS [R12+0x3], R8 
```
which is exactly what happens without explicit coalescing.

Reply:
So let us keep it as it is.